core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -704,7 +705,27 @@ public int hashCode() {
         /**
          * Returns the next operation in the snapshot or <code>null</code> if we reached the end.
          */
-        Translog.Operation next() throws IOException;
+        Position next() throws IOException;
+
+    }
+
+    public static class Position {
```

Comment:
I agree it will be cleaner to have this explicitly exposed in the operation. If we do not implement by reading from the translog location, it means we'll need another integer we can fall back to, but it's potentially not a big deal.

There is another aspect here that I wanted to mention - the version map is used for concurrency control. We don't really need it now  because the translog replays things in a safe order - we never write older version of a doc after new versions at the moment and no one writes into the engine externally. Given how brittle this is and the fact that we do have plans to change these semantics, I'm personally +0 to do it the way Jason did (and fix Operation to report its size directly, based on the underlying translog location). I can also live with not changing the snapshot readers now, the way Simon suggested.